edge-client/reset_factory_settings.c

```diff
@@ -132,6 +139,90 @@ EDGE_LOCAL void rfs_reset_factory_settings_request_cb(void *arg)
     free(arg);
 }
 
+#ifdef RFS_GPIO
+
+#ifdef RFS_GPIO_INVERTED
+#define RFS_GPIO_POLARITY (GPIOD_LINE_EVENT_FALLING_EDGE ^ GPIOD_LINE_EVENT_RISING_EDGE)
+#else
+#define RFS_GPIO_POLARITY 0
+#endif
+
+static void *rfs_gpio_thread(void *arg)
+{
+    struct gpiod_line *line = arg;
+    struct gpiod_line_event last_event = { 0 };
+    last_event.event_type = GPIOD_LINE_EVENT_FALLING_EDGE ^ RFS_GPIO_POLARITY;
+
+    for (;;) {
+        struct gpiod_line_event event;
+        // Wait for and read an event
+        if (gpiod_line_event_read(line, &event)) {
+            tr_err("gpiod_line_event_read error");
+            break;
+        }
+        // If unknown event, ignore
+        if (event.event_type != GPIOD_LINE_EVENT_FALLING_EDGE && event.event_type != GPIOD_LINE_EVENT_RISING_EDGE) {
+            continue;
+        }
+        event.event_type ^= RFS_GPIO_POLARITY;
+        tr_info("rfs button %s", event.event_type == GPIOD_LINE_EVENT_RISING_EDGE ? "pressed", "released");
+        // Look for release after press
+        if (last_event.event_type == GPIOD_LINE_EVENT_RISING_EDGE && event.event_type == GPIOD_LINE_EVENT_FALLING_EDGE) {
+            // Check if it was pressed for at least 2 seconds
+            long long secs = event.ts.tv_sec - last_event.ts.tv_sec;
+            if (event.ts.tv_nsec < last_event.ts.tv_nsec) {
+                secs--;
+            }
+            // If so, request the reset, and exit the GPIO handler
+            if (secs >= 2) {
```

Comment:
Should the # of seconds be also configurable? Should it be 10?

Reply:
Certainly could make it configurable. No idea what the default will be.

Nothing will happen until it's released, so I had a vague feeling that having an excessively-long press requirement might be counter-intuitive. Hold for 5 seconds, and nothing happens, would you try again for 10?

Reply:
2 secs sounds too short for accidental resets. Typically time is ~30sec in consumer devices, but I don't know what kind of requirement this kind of devices has.

Reply:
Changed default to 10, and made it configurable.

Reply:
Excellent!